File: ios/Classes/CallModeListener.swift
import Foundation
import CallKit

class CallModeListener {
    static private let defaultTimerIntervalMicroseconds: Int = 100000

    private var timerIntervalMicroseconds: Int = defaultTimerIntervalMicroseconds
    private var timer: Timer?
    private var callback: ((CallMode)->())?
    private var callMode: CallMode
    
    init() {
        self.callMode = CallModeListener.getCallModeInternal()
    }
    
    public func getCallMode() -> CallMode {
        return CallModeListener.getCallModeInternal()
    }
    
    func watch(callback: @escaping ((CallMode)->())) {
        self.callback = callback
        resetTimer()
    }

    func stopWatcher() {
        self.callback = nil
        stopAndRemoveTimer()
    }

    func setWatcherInterval(intervalInMicroseconds: Int) {
        timerIntervalMicroseconds = intervalInMicroseconds

        resetTimer()
    }

    private func resetTimer() {
        if (callback == nil) { return }

        stopAndRemoveTimer()

        timer = Timer.scheduledTimer(
            withTimeInterval: Double(timerIntervalMicroseconds / 1000000),
            repeats: true
        ) { timer in
            if (self.callback == nil) {
                self.stopAndRemoveTimer()
            }

            let newCallMode = self.getCallMode()
            
            if (self.callMode != newCallMode) {
                self.callMode = newCallMode
                self.callback?(newCallMode)
            }
        }
    }

    private func stopAndRemoveTimer() {
        timer?.invalidate()
        timer = nil
    }
    
    static private func getCallModeInternal() -> CallMode {
        for call in CXCallObserver().calls {
            if !call.hasEnded {
                if call.hasConnected {
                    return CallMode.inProgress
                } else {
                    return CallMode.ringing
                }
            }
        }

        return CallMode.none
    }
}

File: android/src/main/kotlin/com/kardenvan/call_mode/CallModeListener.kt
package com.kardenvan.call_mode

import android.content.Context
import android.media.AudioManager
import java.util.Timer
import kotlin.concurrent.timer

class CallModeListener(private var context: Context) {
    private val defaultTimerIntervalMicroseconds: Int = 100000

    private var intervalInMicroseconds: Int = defaultTimerIntervalMicroseconds
    private var timerTask: Timer? = null
    private var callback: ((CallMode)->Unit)? = null
    private var callMode: CallMode

    init {
        callMode = getCallModeProgress()
    }

    fun getCallModeProgress() : CallMode {
        return getCallModeInternal()
    }

    fun watch(callback: ((CallMode)->Unit)) {
        this.callback = callback
        resetTimer()
    }

    fun stopWatcher() {
        this.callback = null
        stopAndRemoveTimer()
    }

    fun setWatcherInterval(intervalInMicroseconds: Int) {
        this.intervalInMicroseconds = intervalInMicroseconds

        resetTimer()
    }

    private fun resetTimer() {
        if (callback == null) return

        stopAndRemoveTimer()

        val periodInMilliseconds = (intervalInMicroseconds / 1000).toLong()

        timerTask = timer(
            period = periodInMilliseconds,
        ) {
            if (callback == null) {
                stopAndRemoveTimer()
            }

            val newCallMode = getCallModeProgress()

            if (callMode != newCallMode) {
                callMode = newCallMode
                triggerCallback(newCallMode)
            }
        }
    }

    private fun triggerCallback(value: CallMode) {
        callback?.let { it(value) }
    }

    private fun stopAndRemoveTimer() {
        timerTask?.cancel()
        timerTask?.purge()
        timerTask = null
    }

    private fun getCallModeInternal() : CallMode {
        val service: AudioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

        return when (service.mode) {
            AudioManager.MODE_IN_CALL -> CallMode.IN_PROGRESS
            AudioManager.MODE_IN_COMMUNICATION -> CallMode.IN_PROGRESS
            AudioManager.MODE_RINGTONE -> CallMode.RINGING
            else -> CallMode.NONE
        }
    }
}